lib/utils.js

```diff
@@ -29,3 +29,52 @@ exports.shallowObjEqual = function(a, b) {
       a[key] === b[key];
   });
 };
+
+var defaultCompileOptions = {
+  generateLetDeclarations: false,
+  parse: require("./parsers/default.js").parse,
+  ast: false
+};
+
+exports.getOption = function (options, name) {
+  if (options && hasOwn.call(options, name)) {
+    return options[name];
+  }
+  return defaultCompileOptions[name];
+};
+
+exports.getNamesFromPattern = function (pattern) {
+  var queue = [pattern];
+  var names = [];
+
+  for (var i = 0; i < queue.length; ++i) {
+    var pattern = queue[i];
+    if (! pattern) {
+      continue;
+    }
+
+    switch (pattern.type) {
+    case "Identifier":
+      names.push(pattern.name);
+      break;
+    case "Property":
+    case "ObjectProperty":
+      queue.push(pattern.value);
+      break;
+    case "AssignmentPattern":
+      queue.push(pattern.left);
+      break;
+    case "ObjectPattern":
+      queue.push.apply(queue, pattern.properties);
+      break;
+    case "ArrayPattern":
+      queue.push.apply(queue, pattern.elements);
+      break;
+    case "RestElement":
+      queue.push(pattern.argument);
+      break;
+    }
```

Comment:
It's part of the `module.runModuleSetters` wrapping that `AssignmentVisitor` does.

Consider this code:
```js
export let a, rest;

// Elsewhere in the file:
[a, ...rest] = foo;
```

This becomes
```js
module.export({
  a: function() { return a },
  rest: function() { return rest }
});
let a, rest;

// Elsewhere in the file:
module.runModuleSetters([a, ...rest] = foo);
```
Because `a` and `rest` are exported, the `AssignmentVisitor` uses `getNamesFromPattern` to find places where they are reassigned, and wraps those `AssignmentExpression`s with a call to `module.runModuleSetters`.

Now, if we could be sure that the assignment happened before the module finished loading for the first time, then we could avoid the wrapping, because `module.runModuleSetters` is idempotent and we call it automatically when a module finishes loading. In general, though, Reify has to worry about reassignments that happen long after the module finishes loading.

It's tempting to disable the `AssignmentVisitor` when using the top-level parser, but I haven't fully thought through the consequences of that.

It's also worth mentioning that the top-level parsers sacrifices spec compliance by not updating exports that get reassigned in nested scopes. It's worth thinking about whether that's really OK, but it's also entirely possible for the library author (e.g. you) to call `module.runModuleSetters()` manually in cases where updating the exported value is really important. If that's too much to ask, the developer should not use the top-level parser. With clever optimizations comes additional responsibility for correctness.